mantidproject/mantid
# Add measurement files metadata to ISIS Reflectometry ORSO file

### Description of work

#### Summary of work

Adds the incident angle and transmission run metadata to the ISIS ORSO reduced reflectivity data file. This completes part of the first bullet point on issue #36395.

As the comments in the code explain, I haven't used the `FileFinder` to get the filenames here because it runs too slowly for the performance we require from this algorithm. Instead the algorithm will try to construct the run file name based on the run value entered into the history (omitting the file extension, as it is not required). Under normal use scientists will simply enter run numbers into the ISIS Refl GUI and these are not a problem to format correctly. There are some edge cases that may result in incorrect file names being included in the metadata, but I have tried to write the code so that the scientists would have to be doing something very deliberately obscure to encounter these.

I have written tests to cover the new functionality, however I am opening a maintenance issue to take a proper look at a better approach for these (see #36777), as I think there are a number of things we ideally should improve.



Refs #36395. 




### To test:


1) Ask Rachel for a batch file to test with and save it somewhere on your machine. This batch file populates transmission runs on the Experiment Settings tab of the GUI, which is needed for testing.
1) Go to `Interfaces -> Reflectometry -> ISIS Reflectometry`.
1) Load the batch file by going to `Batch -> Load` and selecting the batch file location.
1) Go to the Save ASCII tab. Set a Save Path for saving the file to.
1) In the File Format section, select `ORSO ASCII (*.ort)` from the drop-down.
1) In the Automatic Save section, tick option `Save as ASCII on completion`.
1) Go back to the Runs tab, click on the first group in the main table to select it and click the `Process` button to process the group.
1) The reduction should complete successfully and there should be a successful call to `SaveISISReflectometryORSO` recorded in the messages pane. If you navigate to the save path you set in the Save ASCII tab, you should see a `.ort` file. Open this in a text editor and check the following:
    - There are entries in the header under `measurement -> data_files` for each individual sample run, with a comment identifying the incident angle.
    - There are entries in the header under  `measurement -> additional_files` for each transmission run, with a comment identifying whether it's a first or second transmission run.
The format of all measurement file entries should be `{instrument name}{zero padding}{run number}`.
1) Check some different ways of entering run numbers to ensure the metadata output to the `.ort` file is sensible. As mentioned in the PR main description, there are edge cases where you can end up with incorrect file names in the `.ort` header, but these shouldn't be encountered under normal use.





---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
